File: tp3/main_empty.py
import random

from utils import exercice

#
# EXERCICE 1
#

@exercice
def exercice1():
    # ******************** Votre code ci-dessous ********************
    number = int(input("Entrez un nombre: "))
    print(f"{number} est {'positif' if number>0 else 'nul' if number==0 else 'négatif'}.")
    # ******************** Votre code ci-dessus *********************

@exercice
def exercice2():
    ma_liste = [4, 3, 7, 10, 42, 1]
    ind = 0
    biggest_number = -float('inf')
    # ******************** Votre code ci-dessous ********************
    while ind < len(ma_liste):
        biggest_number = ma_liste[ind] if ma_liste[ind] > biggest_number else biggest_number
        ind += 1
    print(f"{biggest_number} est le nombre le plus grand (selon while)")
    biggest_number = -float('inf')
    for _, element in enumerate(ma_liste):
        biggest_number = element if element > biggest_number else biggest_number
    print(f"{biggest_number} est le nombre le plus grand (selon for enumerate)")

    # ******************** Votre code ci-dessus *********************

@exercice
def exercice3():
    ma_liste = [4, 3, 7, 10, 42, 1]
    # ******************** Votre code ci-dessous ********************
    ind = 0
    smallest_number = float('inf')
    # ******************** Votre code ci-dessous ********************
    while ind < len(ma_liste):
        smallest_number = ma_liste[ind] if ma_liste[ind] < smallest_number else smallest_number
        ind += 1
    print(f"{smallest_number} est le nombre le plus petit (selon while)")
    smallest_number = float('inf')
    for _, element in enumerate(ma_liste):
        smallest_number = element if element < smallest_number else smallest_number
    print(f"{smallest_number} est le nombre le plus petit (selon for enumerate)")
    # ******************** Votre code ci-dessus *********************

@exercice
def exercice4():
    # ******************** Votre code ci-dessous ********************
    hauteur = int(input("Hauteur du carré: "))
    for i in range(hauteur):
        print(" "*(hauteur-i), "*"*(i*2+1))
    for i in range(hauteur-2, -1, -1):
        print(" "*(hauteur-i), "*"*(i*2+1))

    # ******************** Votre code ci-dessus *********************

def fibo_iteratif(n:int):   # fonction à un seul argument
    # TODO n° 2 de l'ex 5:
    # ******************** Votre code ci-dessous ********************
    a, b = 0, 1
    for i in range(0, n):
        a, b = b, a + b
    return a
    # ******************** Votre code ci-dessus *********************

@exercice
def exercice5():
    # ******************** Votre code ci-dessous ********************
    F_n = int(input("Entier à calculer (Fib): "))
    print(f"{fibo_iteratif} returns {fibo_iteratif(F_n)}")
    # ******************** Votre code ci-dessus *********************

@exercice
def exercice6():
    # ******************** Votre code ci-dessous ********************
    liste_1 = ["Alice", "Bob"]
    liste_2 = ["Charlie", "Eve"]
    ma_liste = liste_1+liste_2
    print(f"Ma liste utilisant '+=': {ma_liste}")
    print("\n")

    ma_liste = liste_1
    for element in liste_2:
        ma_liste.append(element)
    print(f"Ma liste utilisant '.append' {ma_liste}")
    # ******************** Votre code ci-dessus *********************

def tri(ma_liste):  # fonction à un seul argument
    # TODO n° 2 de l'exercice 7 (version in-place):
    # ******************** Votre code ci-dessous ********************
    pass
    # ******************** Votre code ci-dessus *********************

@exercice
def exercice7():
    import random
    random.seed(42)
    # Liste pseudo-aléatoire avec seed fixée (la liste ne change pas si vous relancez)
    ma_liste = [random.randrange(100) for _ in range(20)]   # 20 nombres
    # Cette façon de créer des listes s'appelle "list comprehension" et s'exécute
    # généralement plus rapidement qu'une boucle for en plusieurs lignes.
    # Cependant, par soucis de compréhension, il est des fois préférable d'écrire
    # la boucle sur plusieurs lignes.
    # Le _ entre le for et le in est juste pour dire qu'on n'utilise pas cette valeur.
    print("Liste à trier:", ma_liste)
    # ******************** Votre code ci-dessous ********************
    pass
    # ******************** Votre code ci-dessus *********************

@exercice
def exercice8():
    import random
    random.seed(7)
    # Liste pseudo-aléatoire avec seed fixée (la liste ne change pas si vous relancez)
    ma_liste = [random.randrange(100) for _ in range(10)]   # 10 nombres
    print("Liste à séparer:", ma_liste)
    # ******************** Votre code ci-dessous ********************
    pairs = [i for i in ma_liste if i%2 == 0]
    print(f"\nPairs: {pairs}")
    impaires = [i for i in ma_liste if i%2 == 1]
    print(f"\nImpairs: {impaires}")
    # ******************** Votre code ci-dessus *********************

@exercice
def exercice9():
    # ******************** Votre code ci-dessous ********************
    pass
    # ******************** Votre code ci-dessus *********************

@exercice
def exercice10():
    liste_de_liste = [[93, 49, 71], [36, 83, 53], [66, 32, 51]]
    liste_applatie = []
    # ******************** Votre code ci-dessous ********************
    pass
    # ******************** Votre code ci-dessus *********************

# Fonction à compléter ex11:
def est_palindrome(ma_chaine):
    # ******************** Votre code ci-dessous ********************
    pass
    # ******************** Votre code ci-dessus *********************

@exercice
def exercice11():
    ma_chaine = input("Entrez la chaine de caractères: ")
    # ******************** Votre code ci-dessous ********************
    pass
    # ******************** Votre code ci-dessus *********************


if __name__ == "__main__":
    # Astuce : commenter tous les exercices sauf celui en cours pour gagner du temps !
    ##exercice1()
    ##exercice2()
    ##exercice3()
    ##exercice4()
    ##exercice5()
    exercice6()
    exercice7()
    exercice8()
    exercice9()
    exercice10()
    exercice11()

File: tp08/main_empty.c
#include <math.h>
#include <stdio.h>


void exercice1(void){
    printf("\n\nEXERCICE 1\n\n");
    
    char ma_chaine[] = "J'aime les pizzas"; // déclaration et affectation
    
    /******************** Votre code ci-dessous ********************/
    printf("%c", ma_chaine[0]);
    for (int i = 1; ma_chaine[i] != '\0'; i++){
        printf("-%c", ma_chaine[i]);
    }
    printf("\n");
    /******************** Votre code ci-dessus *********************/
    
    return;
}


void exercice2(void){
    printf("\n\nEXERCICE 2\n\n");
    
    char ma_chaine[] = "On est à Genève";
    
    /******************** Votre code ci-dessous ********************/ /*
    printf("Char[20] = %c, Char[1000000] = %c", ma_chaine[20], ma_chaine[1000000]); // THis crashes the code
    */
    /******************** Votre code ci-dessus *********************/
    
    return;
}


void exercice3(void) {
    printf("\n\nEXERCICE 3\n\n");

    int height = 7;
    
    /******************** Votre code ci-dessous ********************/
    for (int i = 1; i <= height; i++){
        for (int j = 0; j < i; j++){
            printf("*");
        }
    printf("\n");
    }
    printf("\n");
    /******************** Votre code ci-dessus *********************/

    return;
}

void exercice4(void) {
    printf("\n\nEXERCICE 4\n\n");

    int x = 42;
    printf("++x: %d\n", ++x);
    printf("x après %d\n\n", x);

    x = 42;
    printf("x++: %d\n", x++);
    printf("x après %d\n\n", x);

    x = 42;
    printf("x+=1: %d\n", x += 1);
    printf("x après %d\n\n", x);

    /******************** Votre code ci-dessous ********************/
    x = 42;
    printf("--x: %d\n", --x);
    printf("x après %d\n\n", x);

    x = 42;
    printf("x--: %d\n", x--);
    printf("x après %d\n\n", x);

    x = 42;
    printf("x-=1: %d\n", x -= 1);
    printf("x après %d\n\n", x);
    /******************** Votre code ci-dessus *********************/
    
    return;
}

/******************** Votre fonction ci-dessous ********************/
int square(int x){
    return x * x;
}
/******************** Votre fonction ci-dessus *********************/

void exercice5(void) {
    printf("\n\nEXERCICE 5\n\n");
    /******************** Votre code ci-dessous ********************/
        printf("%d", square(5));
    /******************** Votre code ci-dessus *********************/

    return;
}


void exercice6(void) {
    printf("\n\nEXERCICE 6\n\n");
    
    /******************** Votre code ci-dessous ********************/
    //printf("3.0: %d, 3.14: %d, hello: %d", square(3.0), square(3.14), square("hello")); //square("hello") doesn't compile
    printf("3.0: %d, 3.14: %d", square(3.0), square(3.14)); // returns 3 ^ 2 both times
    /******************** Votre code ci-dessus *********************/

    return;
}

void exercice7(void) {
    printf("\n\nEXERCICE 7\n\n");
    /*
    print(f"ouch = {3 < 4 < 2}") #prints out False
    */

    int ouch = 3 < 4 < 2; // returns 1 or true
    printf("%d\n", ouch);

    return;
}

void exercice8(void) {
    printf("\n\nEXERCICE 8\n\n");
    
    /**************************************************************/
    /********* Exemple d'overflow du type 'unsigned char' *********/
    /**************************************************************/

    unsigned char a, b, c;

    // Vérifions qu'un 'unsigned char' occupe bien 8 bits
    int size_of_unsigned_char = sizeof(a);
    printf("Le type 'unsigned char' utilise %d octets, c'est-à-dire %d bits.\n", size_of_unsigned_char, 8 * size_of_unsigned_char);

    // Exemple d'overflow
    a = 61;
    b = 213;
    c = a + b;  // Overflow : c ne vaut pas 61 + 213 = 274 mais 274 % 256 = 18
    printf("%d + %d = %d\n", a, b, c);
    
    /*************************************/
    /********* Fin de l'exemple *********/
    /*************************************/


    /******************** Votre code ci-dessous ********************/
    int d, e, f, size_of_int;

    size_of_int = sizeof(d);
    printf("Le type 'int' utilise %d octets, c'est-à-dire %d bits.\n", size_of_int, 8 * size_of_int);

    // Exemple d'overflow
    d = 2147483647;
    e = 2;
    f = d + e;
    
    printf("%d + %d = %d\n", d, e, f);
    /******************** Votre code ci-dessus *********************/

    return;
}


int main(void) {   

    // Astuce : commenter tous les exercices sauf celui en cours pour gagner du temps !

    exercice1();
    exercice2();
    exercice3();
    exercice4();
    exercice5();    
    exercice6();
    exercice7();
    exercice8();

    return 0;
}
